functions/Test-DbaAvailabilityGroup.ps1

```diff
@@ -0,0 +1,103 @@
+
+    .PARAMETER IncludePolicyDetail
+
+        Indicates that this function displays the result of each policy evaluation that it performs. The function returns one object per policy evaluation. Each policy object includes the results of evaluation. This information includes
+        whether the policy passed or not, the policy name, and policy category.
+
+    .PARAMETER InputObject
+
+        Enables piping from Get-DbaAvailabilityGroup
+    
+    .PARAMETER EnableException
+        By default, when something goes wrong we try to catch it, interpret it and give you a friendly warning message.
+        This avoids overwhelming you with "sea of red" exceptions, but is inconvenient because it basically disables advanced scripting.
+        Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .NOTES
+        Tags: Hadr, HA, AG, AvailabilityGroup
+        Author: IJeb Reitsma
+
+        Website: https://dbatools.io
+        Copyright: (c) 2018 by dbatools, licensed under MIT
+        License: MIT https://opensource.org/licenses/MIT
+
+    .LINK
+        https://dbatools.io/Test-DbaAvailabilityGroup
+
+    .EXAMPLE
+
+        PS C:\> Test-DbaAvailabilityGroup -SqlInstance sqlserver-0
+
+        Returns the health state on all the Availability Group(s) found on sqlserver-0.
+
+        PS C:\> Test-DbaAvailabilityGroup -SqlInstance sqlserver-0 -AvailabilityGroup test-ag
+
+        Returns the health state on the Availability Group test-ag on sqlserver-0.
+
+        PS C:\> Test-DbaAvailabilityGroup -SqlInstance sqlserver-0 -AvailabilityGroup test-ag -IncludePolicyDetail
+
+        Returns the health state on the Availability Group test-ag on sqlserver-0. Include detailed results.
+
+        PS C:\> Get-DbaAvailabilityGroup -SqlInstance sqlserver-0 | Test-DbaAvailabilityGroup
+
+        Returns the health state on all the Availability Group(s) found on sqlserver-0.
+
+        PS C:\> Get-DbaAvailabilityGroup -SqlInstance sqlserver-0 -AvailabilityGroup test-ag | Test-DbaAvailabilityGroup
+
+        Returns the health state on the Availability Group test-ag on sqlserver-0.
+
+    #>
+    [CmdletBinding()]
+    param (
+        [DbaInstanceParameter[]]$SqlInstance,
+        [PSCredential]$SqlCredential,
+        [string[]]$AvailabilityGroup,
+        [parameter(ValueFromPipeline)]
+        [Microsoft.SqlServer.Management.Smo.AvailabilityGroup[]]$InputObject,
+        [switch]$IncludeUserPolicy,
+        [switch]$IncludePolicyDetail,
+        [switch]$EnableException
+    )
+    process {
+
+        if ($SqlInstance) {
+            $InputObject += Get-DbaAvailabilityGroup -SqlInstance $SqlInstance -SqlCredential $SqlCredential -AvailabilityGroup $AvailabilityGroup
+        }
+
+        foreach ($ag in $InputObject) {
+            #$results = Test-SqlAvailabilityGroup -Path ($ag.Urn | Convert-UrnToPath) -IncludeUserPolicy:$IncludeUserPolicy -IncludePolicyDetail:$IncludePolicyDetail
```

Comment:
I'm so sorry I didn't mention this earlier, we try not to rely on 3rd party modules. basically, if you can recreate whatever `Test-SqlAvailabilityGroup` does, that would be good. Though there are a number of ways to test an AG. Do you see any room for improvement over what `Test-SqlAvailabilityGroup`  has to offer? (I don't know the answer, I don't use it, but there may be opportunity here).

Reply:
I used this blog post [https://blogs.msdn.microsoft.com/sqlalwayson/2012/02/08/the-alwayson-health-model-part-1-health-model-architecture/](url).

The function should always return the same state information as the dashboard in SSMS (Health/Error/Warning). Otherwise it's still necessary to check the dashboard.

The health state is determined with a set of PBM policies. Dbatools does not have a function to evaluate a PBM policy (I think this would also need the sqlserver module).
I think smo or T-SQL cannot be used.

I do not see another option.